sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/WebExceptionRetryPolicyTest.java

```diff
@@ -0,0 +1,338 @@
+
+        assertThat(dsr.getResponseTimeout()).isEqualTo(Duration.ofSeconds(10));
+        validateSuccess(shouldRetry, ShouldRetryValidator.builder().
+            nullException().
+            shouldRetry(true).
+            backOffTime(Duration.ofSeconds(0)).
+            build());
+
+        // 4th Attempt - retry is set to false, as we only make 3 retry attempts for now.
+        retryContext.addStatusAndSubStatusCode(408, 10002);
+        webExceptionRetryPolicy.onBeforeSendRequest(dsr);
+        shouldRetry = webExceptionRetryPolicy.shouldRetry(cosmosException);
+
+        validateSuccess(shouldRetry, ShouldRetryValidator.builder().
+            nullException().
+            shouldRetry(false).
+            build());
+    }
+
+    @Test(groups = "unit")
+    public void shouldNotRetryOnTimeoutForWriteOperations() throws Exception {
+        GlobalEndpointManager endpointManager = Mockito.mock(GlobalEndpointManager.class);
+        Mockito.doReturn(new URI("http://localhost:")).when(endpointManager).resolveServiceEndpoint(Mockito.any(RxDocumentServiceRequest.class));
+        Mockito.doReturn(Mono.empty()).when(endpointManager).refreshLocationAsync(Mockito.eq(null), Mockito.eq(true));
+
+
+        Exception exception = ReadTimeoutException.INSTANCE;
+        CosmosException cosmosException = BridgeInternal.createCosmosException(null, HttpConstants.StatusCodes.REQUEST_TIMEOUT, exception);
+        BridgeInternal.setSubStatusCode(cosmosException, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_READ_TIMEOUT);
+
+        RxDocumentServiceRequest dsr;
+        Mono<ShouldRetryResult> shouldRetry;
+
+        //Data Plane Write - Should not retry
+        dsr = RxDocumentServiceRequest.createFromName(mockDiagnosticsClientContext(),
+            OperationType.Create, "/dbs/db/colls/col/docs/doc", ResourceType.Document);
+
+        WebExceptionRetryPolicy webExceptionRetryPolicy = new WebExceptionRetryPolicy(new RetryContext());
+        webExceptionRetryPolicy.onBeforeSendRequest(dsr);
+        shouldRetry = webExceptionRetryPolicy.shouldRetry(cosmosException);
+
+        validateSuccess(shouldRetry, ShouldRetryValidator.builder()
+            .nullException()
+            .shouldRetry(false)
+            .build());
+
+        //Metadata Write - Should not Retry
+        dsr = RxDocumentServiceRequest.createFromName(mockDiagnosticsClientContext(),
+            OperationType.Create, "/dbs/db/colls/col/docs/docId", ResourceType.Database);
+
+        webExceptionRetryPolicy = new WebExceptionRetryPolicy(new RetryContext());
+        webExceptionRetryPolicy.onBeforeSendRequest(dsr);
+        shouldRetry = webExceptionRetryPolicy.shouldRetry(cosmosException);
+
+        validateSuccess(shouldRetry, ShouldRetryValidator.builder()
+            .nullException()
+            .shouldRetry(false)
+            .build());
+    }
+
+    @Test(groups = "unit")
+    public void httpNetworkFailureOnAddressRefresh() throws Exception {
```

Comment:
Can we do a test on the behavior of address refresh retries when the SDK gets `GATEWAY_ENDPOINT_UNAVAILABLE`? This could also materialize in network issues with the gateway.

Reply:
I will send in another PR for handling the 503 errors. Will add the test case in that PR.